Advanced Template documentation (#2645)
tinymce/tinymce-docs: modules/ROOT/pages/advanced-templates.adoc

= Advanced Template plugin
:navtitle: Advanced Template
:description: Create and edit complex templates in {productname} for use with {productname}.
:description_short: Create templates.
:keywords: plugin, advanced template, templates, advtemplate.
:pluginname: Advanced Template
:plugincode: advtemplate
:plugincategory: premium

include::partial$misc/admon-premium-plugin.adoc[]

> include::partial$misc/admon-requires-6.4v.adoc[]
> 

Reviewer: ```suggestion
include::partial$misc/admon-requires-6.4v.adoc[]

NOTE: The documentation for this page is a work in progress. Further details on the {pluginname} plugin will be added to this page soon.
```
Might be important to forewarn that this page is a WIP, and to check back later for more information

> The Advanced Template Premium plugin enables the creation of complex, interactive templates using {productname}.

Reviewer: ```suggestion
The {pluginname} Premium plugin enables the creation of complex, interactive templates using {productname}.
```
Might be a good idea to grep this, in case we change the plugin name.

They can also be added to a {productname} instance by end-users selecting content in a {productname} document and saving the selection as a template via the {productname} user-interface.

IMPORTANT: This initial Advanced Template release does not support template editing. To update a particular template; insert the template into an editor; make edits; select the, now edited, content; and then create a new template (using, for example, the *Tools > Save as Template…* menu item).

////
== Interactive example

liveDemo::{plugincode}[]
////

== Basic setup

To setup the {pluginname} plugin user-interface in the editor

* add `{plugincode}` to the `plugins` option in the editor configuration;

> * add `inserttemplate` and `addtemplate` to the `toolbar` option in the editor configuration;
> 

Reviewer: ```suggestion
* add `inserttemplate` and `addtemplate` to the `toolbar` option in the editor configuration;

* add each Advanced Template option to the editor configuration;
```

For example:

> [source,js]
> ----
> tinymce.init({
>   selector: "textarea",
>   plugins: "advtemplate",
>   toolbar: "inserttemplate addtemplate",
>   advtemplate_list: () => {
>     return Promise.resolve([
>     {
>       id: '1',
>       title: 'Template 1'
>     },
>     {
>       id: '2',
>       title: 'Category 2',
>       items: [
>       {
>         id: '3',
>         title: 'Template 2.1',
>       },
>       {
>         id: '4',
>         title: 'Template 2.3',
>       }
>     ]
>   }
>   ]);
>   }
> });
> ----

Reviewer: ```suggestion
[source,js]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  advtemplate_list: () => {
    return Promise.resolve([
      {
        id: '1',
        title: 'Resolving tickets',
        content: '<p>As we have not heard back from you in over a week, so we have gone ahead and resolved your ticket</p>'
      },
      {
        id: '2',
        title: 'Quick replies',
        items: [
          {
            id: '3',
            title: 'Message received',
            content: '<p>Just a quick note to say we have received your message, and will get back to you within 48 hours.</p>'
          },
          {
            id: '4',
            title: 'Progress update',
            content: '</p>Just a quick note to let you know we are still working on your case</p>'
          }
        ]
      }
    ]);
  }
});
----
```
Fixing the indentation and content here too

= Advanced Template plugin
:navtitle: Advanced Template
:description: Create and edit complex templates in {productname} for use with {productname}.
:description_short: Create templates.
:keywords: plugin, advanced template, templates, advtemplate.
:pluginname: Advanced Template
:plugincode: advtemplate
:plugincategory: premium

include::partial$misc/admon-premium-plugin.adoc[]

include::partial$misc/admon-requires-6.4v.adoc[]

> NOTE: The documentation for this page is a work in progress. Further details on the {pluginname} plugin will be added to this page soon.
> The {pluginname} Premium plugin enables the creation of complex, interactive templates using {productname}.
> 
> They can also be added to a {productname} instance by end-users selecting content in a {productname} document and saving the selection as a template via the {productname} user-interface.
> 
> IMPORTANT: This initial {pluginname} release does not support template editing. To update a particular template; insert the template into an editor; make edits; select the, now edited, content; and then create a new template (using, for example, the *Tools > Save as Template…* menu item).
> 
> ////
> == Interactive example
> 
> liveDemo::{plugincode}[]
> ////
> 
> == Basic setup
> 
> To setup the {pluginname} plugin user-interface in the editor
> 
> * add `{plugincode}` to the `plugins` option in the editor configuration;
> 
> * add `inserttemplate` and `addtemplate` to the `toolbar` option in the editor configuration;
> 
> * add each {pluginname} option to the editor configuration;
> For example:
> 

Reviewer: ```suggestion
NOTE: The documentation for this page is a work in progress. Further details on the {pluginname} plugin will be added to this page soon.

The {pluginname} Premium plugin enables the creation of complex, interactive templates using {productname}.

They can also be added to a {productname} instance by end-users selecting content in a {productname} document and saving the selection as a template via the {productname} user-interface.

IMPORTANT: This initial {pluginname} release does not support template editing. To update a particular template; insert the template into an editor; make edits; select the, now edited, content; and then create a new template (using, for example, the *Tools > Save as Template…* menu item).

////
== Interactive example

liveDemo::{plugincode}[]
////

== Basic setup

To setup the {pluginname} plugin user-interface in the editor

* add `{plugincode}` to the `plugins` option in the editor configuration;

* add `inserttemplate` and `addtemplate` to the `toolbar` option in the editor configuration;

* add each {pluginname} option to the editor configuration;

For example:

```
Line spacing

////

== Basic setup

To setup the {pluginname} plugin user-interface in the editor

* add `{plugincode}` to the `plugins` option in the editor configuration;

* add `inserttemplate` and `addtemplate` to the `toolbar` option in the editor configuration;

* add each {pluginname} option to the editor configuration;

For example:

[source,js]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  advtemplate_list: () => {
    return Promise.resolve([
      {
        id: '1',
        title: 'Resolving tickets',
        content: '<p>As we have not heard back from you in over a week, so we have gone ahead and resolved your ticket</p>'
      },
      {
        id: '2',
        title: 'Quick replies',
        items: [
          {
            id: '3',
            title: 'Message received',
            content: '<p>Just a quick note to say we have received your message, and will get back to you within 48 hours.</p>'
          },
          {
            id: '4',
            title: 'Progress update',
            content: '</p>Just a quick note to let you know we are still working on your case</p>'
          }
        ]
      }
    ]);
  }
});
----

> NOTE: The above example only demonstrates the {pluginname} user interface. An xref:#advtemplate_list[`advtemplate_list`] option has been added, with an included function returning a promise that contains a basic Template structure. This allows the {pluginname} user interface to load and present in a {productname} editor instance. This basic setup example cannot create, edit, or save new {pluginname} items.

Reviewer 2: This plugin configured according to the above example will not be able to preview template content

Author: > The plugin configured according to the above example will not be able to preview and insert templates' content

Is there any way to set up the promise such that, should someone copy-and-paste the sample into a fiddle, the UI would also preview the selected template and allow the template text be inserted into the TinyMCE editor instance?

Author: I’ve pushed an [edit to this sentence](https://github.com/tinymce/tinymce-docs/commit/2cd65418d1645dae6b9dd0f5722a3d39caf0a447) that explicitly acknowledges the limitation noted by @vpyshnenko above.

= Advanced Template plugin
:navtitle: Advanced Template
:description: Create and edit complex templates in {productname} for use with {productname}.
:description_short: Create and edit templates.
:keywords: plugin, advanced template, templates, advtemplate.
:pluginname: Advanced Template
:plugincode: advtemplate
:plugincategory: premium

include::partial$misc/admon-premium-plugin.adoc[]

include::partial$misc/admon-requires-6.4v.adoc[]

> The Advanced Template Premium plugin enables the creation and editing of complex, interactive templates using {productname}.

Reviewer 2: Unfortunately, we do not currently support template **editing**. Once a template has been created, it can only be previewed. To update the content of the template, users will need to insert the template into the editor content, make edits, select the appropriate region, and then create a new template.

Author: To quote my commit message:

> Edited out references to editing templates.
> 
> Added IMPORTANT admonition outlining using Save as Template… command as editing template workaround.

> Templates are declared as `const` variables that can be setup as part of a {productname} instance.

Author: To quote my commit message:

> Re-wrote the basic setup introduction and example.

== Interactive example

liveDemo::{plugincode}[]

== Basic setup

To setup the {pluginname} plugin in the editor

* add `{plugincode}` to the `plugins` option in the editor configuration;

* add `inserttemplate` and `addtemplate` to the `toolbar` option in the editor configuration;

* add the templates to use in the editor instance as a function returning the Promise which contains the editor instance’s initial template list.

For example:

[source,js]
----
tinymce.init({
    selector: "textarea",
    plugins: "advtemplate",
    toolbar: "inserttemplate addtemplate",
  	advtemplate_list: () => {
      return Promise.resolve([
      {
        id: '1',
        title: 'Template 1'
      },
      {
        id: '2',
        title: 'Category 2',
        items: [
        {
          id: '3',
          title: 'Template 2.1',
        },
        {
          id: '4',
          title: 'Template 2.3',
        }
      ]
    }
  ]);
  }

> });

Reviewer 3: ```suggestion
tinymce.init({
  selector: "textarea",
  plugins: "advtemplate",
  toolbar: "inserttemplate addtemplate",
  advtemplate_list: () => {
    return Promise.resolve([
      {
        id: '1',
        title: 'Template 1'
      },
      {
        id: '2',
        title: 'Category 2',
        items: [
          {
            id: '3',
            title: 'Template 2.1',
          },
          {
            id: '4',
            title: 'Template 2.3',
          }
        ]
      }
    ]);
  }
});
```

= Advanced Template plugin
:navtitle: Advanced Template
:description: Create and edit complex templates in {productname} for use with {productname}.
:description_short: Create templates.
:keywords: plugin, advanced template, templates, advtemplate.
:pluginname: Advanced Template
:plugincode: advtemplate
:plugincategory: premium

include::partial$misc/admon-premium-plugin.adoc[]

include::partial$misc/admon-requires-6.4v.adoc[]

The Advanced Template Premium plugin enables the creation of complex, interactive templates using {productname}.

> Templates are declared as `const` variables that can be setup as part of a {productname} instance.

Reviewer: ```suggestion
Templates are declared as `const` variables that can be setup as part of a {productname} instance.
```
This needs to be removed since we aren't documenting the option any more

Author: Removed.

They can also be added to a {productname} instance by end-users selecting content in a {productname} document and saving the selection as a template via the {productname} user-interface.

> IMPORTANT: This initial Advanced Template release does not support template editing. To update a particular template; insert the template into an editor; make edits; select the, now edited, content; and then create a new template (using, for example, the *Tools > Save as Template…* menu item).

Reviewer 3: ```suggestion
IMPORTANT: This initial Advanced Template release does not support template editing. To update a particular template; insert the template into an editor; make edits; select the now edited content; and then create a new template (using, for example, the *Tools > Save as Template…* menu item).
```

== Interactive example

liveDemo::{plugincode}[]

== Basic setup

To setup the {pluginname} plugin in the editor

* add `{plugincode}` to the `plugins` option in the editor configuration;

* add `inserttemplate` and `addtemplate` to the `toolbar` option in the editor configuration;

> * add the templates to use in the editor instance as a function returning the Promise which contains the editor instance’s initial template list.

Reviewer: ```suggestion
```
Remove this line about using the option

For example:

[source,js]
----
tinymce.init({
    selector: "textarea",
    plugins: "advtemplate",
    toolbar: "inserttemplate addtemplate",

>   	advtemplate_list: () => {
>       return Promise.resolve([
>         {
>           id: '1',
>           title: 'Template 1'
>         },
>         {
>           id: '2',
>           title: 'Category 2',
>           items: [
>             {
>               id: '3',
>               title: 'Template 2.1',
>             },
>             {
>               id: '4',
>               title: 'Category 2.2',
>               items: []
>             }
>           ]
>         }
>     ]);
>   }

Reviewer: ```suggestion
```
This option needs to be removed

Reviewer 2: Use please, this [example](https://www.notion.so/tinycloud/Post-implementation-plugin-details-for-Doc-writers-8442743100f5426d8b7357d2e2eff428?pvs=4#c889f03cd5e843edbf450e537a2f705c) as basic setup.

Reviewer 2: And overall options description should be similar to the one in the [Comments plugin documentation](https://www.tiny.cloud/docs/tinymce/6/comments-callback-mode/#options). Also, look the [Post-implementation plugin details for Doc writers](https://www.notion.so/tinycloud/Post-implementation-plugin-details-for-Doc-writers-8442743100f5426d8b7357d2e2eff428) for the plugin details

});
----

> NOTE: A unique id value must be provided for every input item.

Reviewer: ```suggestion
```
This needs to be removed

> == Options
> 
> The following configuration options affect the behavior of the {pluginname} plugin.

Reviewer: ```suggestion
```
Since we aren't documenting the option, this needs to be removed
